### DISCUSSION TITLE
Doubt about interleaving Server and Client Components

### DISCUSSION THREAD

**Discussion by Colo-Codes** [vercel/next.js#73292] [Help]
### Summary

In the [docs](https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#interleaving-server-and-client-components) you can find this statement:

> Since Client Components are rendered after Server Components, you cannot import a Server Component into a Client Component module (since it would require a new request back to the server). Instead, you can pass a Server Component as props to a Client Component. See the [unsupported pattern](https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#unsupported-pattern-importing-server-components-into-client-components) and [supported pattern](https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#supported-pattern-passing-server-components-to-client-components-as-props) sections below.

I might be misunderstanding this, but I think that the statement is not entirely correct.

When you **import** a component into a Client Component, the imported component is included in the client-side JavaScript bundle. This means:

- The component will run in the browser.
- It must adhere to client-side constraints (no access to server-only APIs).
- Any server-only code inside it will cause errors.

**Example:**

```jsx
// app/ClientComponent.js

'use client';

import ChildComponent from './ChildComponent';

export default function ClientComponent() {
  return (
    <div>
      <ChildComponent />
    </div>
  );
}
```

In this example:

- `ClientComponent` is a Client Component because of the `'use client'` directive.
- It **imports** `ChildComponent`.
- As a result, `ChildComponent` becomes part of the client bundle and is treated as a Client Component, even if it was originally designed as a Server Component.
- If `ChildComponent` contains server-only code (like database queries), it will cause errors when the client tries to execute it.

A Server Component can be imported into a Client Component as long as it doesn't have any code that requires a server to run. When you do this, the Server Component effectively becomes a Client Component, and it's included in the client-side JavaScript bundle. This is acceptable and won't cause issues as long as the component is compatible with the client environment, but it will increase the client-side bundle size.

So, to sum up, you can import both _client components_ and _server components_ into a _client component_, provided that the server component does not have server-only code. I run a local test by importing a server component into a client component and Next.js didn't complain.

![image](https://github.com/user-attachments/assets/3c5876f8-e74b-4350-b6ec-b88b335cfbd1)

Am I missing something here?

**Comment by icyJoseph** [comment]
Yeah, maybe some additional tweaking is needed, because as you say, a component can be opted into the being a client component. They call that unsupported pattern, because, it is possible for that component to eventually, as you say, do server side stuff, that it wouldn't be able to do, had it been opted into the client, and then your code will break...
